Add some simple Playwright tests (#2421)
opensafely-core/opencodelists: docs/adr/0003-browser-testing.md

# Automated browser testing

Date: 2025-05

## Status

Draft

## Context

> While we have unit tests for parts of our code, both frontend and
> backend, user flows throughout the application end-to-end, integrating
> these are not tested.

Reviewer: I think the first paragraph is a bit garbled and although I get the gist I'm not sure I follow it grammatically.

```suggestion
While we have unit tests covering parts of our frontend and
backend code, we lack end-to-end tests for user flows across
the application.
```
- Should we say that we don't currently enforce any Coverage metric, so the tests we do have are definitely not comprehensive?
- Can we give some examples of what a "user flow" is? Registering, viewing an existing codelist, creating a minimal codelist with the builder?
- Can we say anything about why Playwright was chosen over alternatives like Selenium or Cypress?

> It would be useful to have such tests to validate that key user journeys
> continue to function as expected.

Reviewer: ```suggestion
It would be useful to have such tests to validate that key user journeys
continue to function as expected when making changes to the code.
```

## Decision

We will add automated browser tests that are run using Playwright and
pytest-django's `live_server` fixture.

We will use the existing `universe` fixtures for the time being to setup
the coding system data where appropriate, rather than reimplementing
those. It is possible that the existing test fixture system is redesigned in
future to simplify it, and that simplification may help setup of these

> tests.

Reviewer: Can we very briefly state what this is and why that causes a complication/workaround?

These tests will be run separately from the existing unit tests. This is
partly a design choice — to keep the existing test suite fast — and
partly a consequence of the fixture setup — pytest-django's
`live_server` fixture resets database state between tests.

> ## Consequences

Reviewer: Can we state that the consequence of reusing the universe fixture is the need for the workaround to change the fixture scope for these new tests, and why?

Reviewer: Worth stating the positive consequence (why we are doing this): we will have greater assurance there aren't unintended consequences to changes we make?

> The tests will need to be maintained, and new tests added if we add

Reviewer: Also adding additional tests to cover additional user journeys these initial tests don't cover? Should it be something we pursue stand-alone or just in the context of other tickets? How can we encourage that activity?
